File: src/extras-kio/libkacl.cpp
#include <KACL>
#include <QList>
#include <QPair>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <kacl.h>
#include "libkacl.h"
#include "libkacl.hxx"

KACL* KACL_new(const libqt_string aclString) {
    QString aclString_QString = QString::fromUtf8(aclString.data, aclString.len);
    return new VirtualKACL(aclString_QString);
}

KACL* KACL_new2(const KACL* rhs) {
    return new VirtualKACL(*rhs);
}

KACL* KACL_new3(mode_t basicPermissions) {
    return new VirtualKACL(basicPermissions);
}

KACL* KACL_new4() {
    return new VirtualKACL();
}

void KACL_OperatorAssign(KACL* self, const KACL* rhs) {
    self->operator=(*rhs);
}

bool KACL_OperatorEqual(const KACL* self, const KACL* rhs) {
    return (*self == *rhs);
}

bool KACL_OperatorNotEqual(const KACL* self, const KACL* rhs) {
    return (*self != *rhs);
}

bool KACL_IsValid(const KACL* self) {
    return self->isValid();
}

uint16_t KACL_OwnerPermissions(const KACL* self) {
    return self->ownerPermissions();
}

bool KACL_SetOwnerPermissions(KACL* self, uint16_t ownerPermissions) {
    return self->setOwnerPermissions(static_cast<unsigned short>(ownerPermissions));
}

uint16_t KACL_OwningGroupPermissions(const KACL* self) {
    return self->owningGroupPermissions();
}

bool KACL_SetOwningGroupPermissions(KACL* self, uint16_t owningGroupPermissions) {
    return self->setOwningGroupPermissions(static_cast<unsigned short>(owningGroupPermissions));
}

uint16_t KACL_OthersPermissions(const KACL* self) {
    return self->othersPermissions();
}

bool KACL_SetOthersPermissions(KACL* self, uint16_t othersPermissions) {
    return self->setOthersPermissions(static_cast<unsigned short>(othersPermissions));
}

mode_t KACL_BasePermissions(const KACL* self) {
    return self->basePermissions();
}

bool KACL_IsExtended(const KACL* self) {
    return self->isExtended();
}

uint16_t KACL_MaskPermissions(const KACL* self, bool* exists) {
    return self->maskPermissions(*exists);
}

bool KACL_SetMaskPermissions(KACL* self, uint16_t maskPermissions) {
    return self->setMaskPermissions(static_cast<unsigned short>(maskPermissions));
}

uint16_t KACL_NamedUserPermissions(const KACL* self, const libqt_string name, bool* exists) {
    QString name_QString = QString::fromUtf8(name.data, name.len);
    return self->namedUserPermissions(name_QString, exists);
}

bool KACL_SetNamedUserPermissions(KACL* self, const libqt_string name, uint16_t param2) {
    QString name_QString = QString::fromUtf8(name.data, name.len);
    return self->setNamedUserPermissions(name_QString, static_cast<unsigned short>(param2));
}

libqt_list /* of libqt_pair  tuple of libqt_string and uint16_t  */ KACL_AllUserPermissions(const KACL* self) {
    QList<QPair<QString, unsigned short>> _ret = self->allUserPermissions();
    // Convert QList<> from C++ memory to manually-managed C memory
    libqt_pair /* tuple of libqt_string and uint16_t */* _arr = static_cast<libqt_pair /* tuple of libqt_string and uint16_t */*>(malloc(sizeof(libqt_pair /* tuple of libqt_string and uint16_t */) * (_ret.size() + 1)));
    for (qsizetype i = 0; i < _ret.size(); ++i) {
        QPair<QString, unsigned short> _lv_ret = _ret[i];
        // Convert QPair<> from C++ memory to manually-managed C memory
        libqt_string* _lv_first = static_cast<libqt_string*>(malloc(sizeof(libqt_string)));
        uint16_t* _lv_second = static_cast<uint16_t*>(malloc(sizeof(uint16_t)));
        QString _lv_first_ret = _lv_ret.first;
        // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
        QByteArray _lv_first_b = _lv_first_ret.toUtf8();
        libqt_string _lv_first_str;
        _lv_first_str.len = _lv_first_b.length();
        _lv_first_str.data = static_cast<const char*>(malloc(_lv_first_str.len + 1));
        memcpy((void*)_lv_first_str.data, _lv_first_b.data(), _lv_first_str.len);
        ((char*)_lv_first_str.data)[_lv_first_str.len] = '\0';
        *_lv_first = _lv_first_str;
        *_lv_second = _lv_ret.second;
        libqt_pair _lv_out;
        _lv_out.first = static_cast<void*>(_lv_first);
        _lv_out.second = static_cast<void*>(_lv_second);
        _arr[i] = _lv_out;
    }
    libqt_list _out;
    _out.len = _ret.size();
    _out.data = static_cast<void*>(_arr);
    return _out;
}

bool KACL_SetAllUserPermissions(KACL* self, const libqt_list /* of libqt_pair  tuple of libqt_string and uint16_t  */ list) {
    QList<QPair<QString, unsigned short>> list_QList;
    list_QList.reserve(list.len);
    libqt_pair /* tuple of libqt_string and uint16_t */* list_arr = static_cast<libqt_pair /* tuple of libqt_string and uint16_t */*>(list.data);
    for (size_t i = 0; i < list.len; ++i) {
        QPair<QString, unsigned short> list_arr_i_QPair;
        libqt_string* list_arr_i_first = static_cast<libqt_string*>(list_arr[i].first);
        uint16_t* list_arr_i_second = static_cast<uint16_t*>(list_arr[i].second);
        QString list_arr_i_first_0_QString = QString::fromUtf8(list_arr_i_first[0].data, list_arr_i_first[0].len);
        list_arr_i_QPair.first = list_arr_i_first_0_QString;
        list_arr_i_QPair.second = static_cast<unsigned short>(list_arr_i_second[0]);
        list_QList.push_back(list_arr_i_QPair);
    }
    return self->setAllUserPermissions(list_QList);
}

uint16_t KACL_NamedGroupPermissions(const KACL* self, const libqt_string name, bool* exists) {
    QString name_QString = QString::fromUtf8(name.data, name.len);
    return self->namedGroupPermissions(name_QString, exists);
}

bool KACL_SetNamedGroupPermissions(KACL* self, const libqt_string name, uint16_t param2) {
    QString name_QString = QString::fromUtf8(name.data, name.len);
    return self->setNamedGroupPermissions(name_QString, static_cast<unsigned short>(param2));
}

libqt_list /* of libqt_pair  tuple of libqt_string and uint16_t  */ KACL_AllGroupPermissions(const KACL* self) {
    QList<QPair<QString, unsigned short>> _ret = self->allGroupPermissions();
    // Convert QList<> from C++ memory to manually-managed C memory
    libqt_pair /* tuple of libqt_string and uint16_t */* _arr = static_cast<libqt_pair /* tuple of libqt_string and uint16_t */*>(malloc(sizeof(libqt_pair /* tuple of libqt_string and uint16_t */) * (_ret.size() + 1)));
    for (qsizetype i = 0; i < _ret.size(); ++i) {
        QPair<QString, unsigned short> _lv_ret = _ret[i];
        // Convert QPair<> from C++ memory to manually-managed C memory
        libqt_string* _lv_first = static_cast<libqt_string*>(malloc(sizeof(libqt_string)));
        uint16_t* _lv_second = static_cast<uint16_t*>(malloc(sizeof(uint16_t)));
        QString _lv_first_ret = _lv_ret.first;
        // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
        QByteArray _lv_first_b = _lv_first_ret.toUtf8();
        libqt_string _lv_first_str;
        _lv_first_str.len = _lv_first_b.length();
        _lv_first_str.data = static_cast<const char*>(malloc(_lv_first_str.len + 1));
        memcpy((void*)_lv_first_str.data, _lv_first_b.data(), _lv_first_str.len);
        ((char*)_lv_first_str.data)[_lv_first_str.len] = '\0';
        *_lv_first = _lv_first_str;
        *_lv_second = _lv_ret.second;
        libqt_pair _lv_out;
        _lv_out.first = static_cast<void*>(_lv_first);
        _lv_out.second = static_cast<void*>(_lv_second);
        _arr[i] = _lv_out;
    }
    libqt_list _out;
    _out.len = _ret.size();
    _out.data = static_cast<void*>(_arr);
    return _out;
}

bool KACL_SetAllGroupPermissions(KACL* self, const libqt_list /* of libqt_pair  tuple of libqt_string and uint16_t  */ allGroupPermissions) {
    QList<QPair<QString, unsigned short>> allGroupPermissions_QList;
    allGroupPermissions_QList.reserve(allGroupPermissions.len);
    libqt_pair /* tuple of libqt_string and uint16_t */* allGroupPermissions_arr = static_cast<libqt_pair /* tuple of libqt_string and uint16_t */*>(allGroupPermissions.data);
    for (size_t i = 0; i < allGroupPermissions.len; ++i) {
        QPair<QString, unsigned short> allGroupPermissions_arr_i_QPair;
        libqt_string* allGroupPermissions_arr_i_first = static_cast<libqt_string*>(allGroupPermissions_arr[i].first);
        uint16_t* allGroupPermissions_arr_i_second = static_cast<uint16_t*>(allGroupPermissions_arr[i].second);
        QString allGroupPermissions_arr_i_first_0_QString = QString::fromUtf8(allGroupPermissions_arr_i_first[0].data, allGroupPermissions_arr_i_first[0].len);
        allGroupPermissions_arr_i_QPair.first = allGroupPermissions_arr_i_first_0_QString;
        allGroupPermissions_arr_i_QPair.second = static_cast<unsigned short>(allGroupPermissions_arr_i_second[0]);
        allGroupPermissions_QList.push_back(allGroupPermissions_arr_i_QPair);
    }
    return self->setAllGroupPermissions(allGroupPermissions_QList);
}

bool KACL_SetACL(KACL* self, const libqt_string aclStr) {
    QString aclStr_QString = QString::fromUtf8(aclStr.data, aclStr.len);
    return self->setACL(aclStr_QString);
}

libqt_string KACL_AsString(const KACL* self) {
    QString _ret = self->asString();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

void KACL_VirtualHook(KACL* self, int id, void* data) {
    auto* vkacl = dynamic_cast<VirtualKACL*>(self);
    if (vkacl && vkacl->isVirtualKACL) {
        vkacl->virtual_hook(static_cast<int>(id), data);
    }
}

// Base class handler implementation
void KACL_QBaseVirtualHook(KACL* self, int id, void* data) {
    auto* vkacl = dynamic_cast<VirtualKACL*>(self);
    if (vkacl && vkacl->isVirtualKACL) {
        vkacl->setKACL_VirtualHook_IsBase(true);
        vkacl->virtual_hook(static_cast<int>(id), data);
    } else {
        ((VirtualKACL*)self)->virtual_hook(static_cast<int>(id), data);
    }
}

// Auxiliary method to allow providing re-implementation
void KACL_OnVirtualHook(KACL* self, intptr_t slot) {
    auto* vkacl = dynamic_cast<VirtualKACL*>(self);
    if (vkacl && vkacl->isVirtualKACL) {
        vkacl->setKACL_VirtualHook_Callback(reinterpret_cast<VirtualKACL::KACL_VirtualHook_Callback>(slot));
    }
}

void KACL_Delete(KACL* self) {
    delete self;
}

File: include_original/extras-kio/libkacl.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");
const std = @import("std");
pub const struct_constu8_u16 = extern struct { first: []const u8, second: u16 };

/// https://api.kde.org/kacl.html
pub const kacl = struct {
    /// New constructs a new KACL object.
    ///
    /// ``` aclString: []const u8 ```
    pub fn New(aclString: []const u8) QtC.KACL {
        const aclString_str = qtc.libqt_string{
            .len = aclString.len,
            .data = aclString.ptr,
        };

        return qtc.KACL_new(aclString_str);
    }

    /// New2 constructs a new KACL object.
    ///
    /// ``` rhs: QtC.KACL ```
    pub fn New2(rhs: ?*anyopaque) QtC.KACL {
        return qtc.KACL_new2(@ptrCast(rhs));
    }

    /// New3 constructs a new KACL object.
    ///
    /// ``` basicPermissions: u32 ```
    pub fn New3(basicPermissions: u32) QtC.KACL {
        return qtc.KACL_new3(basicPermissions);
    }

    /// New4 constructs a new KACL object.
    ///
    ///
    pub fn New4() QtC.KACL {
        return qtc.KACL_new4();
    }

    /// [Qt documentation](https://api.kde.org/kacl.html#operator-eq)
    ///
    /// ``` self: QtC.KACL, rhs: QtC.KACL ```
    pub fn OperatorAssign(self: ?*anyopaque, rhs: ?*anyopaque) void {
        qtc.KACL_OperatorAssign(@ptrCast(self), @ptrCast(rhs));
    }

    /// [Qt documentation](https://api.kde.org/kacl.html#operator-eq-eq)
    ///
    /// ``` self: QtC.KACL, rhs: QtC.KACL ```
    pub fn OperatorEqual(self: ?*anyopaque, rhs: ?*anyopaque) bool {
        return qtc.KACL_OperatorEqual(@ptrCast(self), @ptrCast(rhs));
    }

    /// [Qt documentation](https://api.kde.org/kacl.html#operator-not-eq)
    ///
    /// ``` self: QtC.KACL, rhs: QtC.KACL ```
    pub fn OperatorNotEqual(self: ?*anyopaque, rhs: ?*anyopaque) bool {
        return qtc.KACL_OperatorNotEqual(@ptrCast(self), @ptrCast(rhs));
    }

    /// [Qt documentation](https://api.kde.org/kacl.html#isValid)
    ///
    /// ``` self: QtC.KACL ```
    pub fn IsValid(self: ?*anyopaque) bool {
        return qtc.KACL_IsValid(@ptrCast(self));
    }

    /// [Qt documentation](https://api.kde.org/kacl.html#ownerPermissions)
    ///
    /// ``` self: QtC.KACL ```
    pub fn OwnerPermissions(self: ?*anyopaque) u16 {
        return qtc.KACL_OwnerPermissions(@ptrCast(self));
    }

    /// [Qt documentation](https://api.kde.org/kacl.html#setOwnerPermissions)
    ///
    /// ``` self: QtC.KACL, ownerPermissions: u16 ```
    pub fn SetOwnerPermissions(self: ?*anyopaque, ownerPermissions: u16) bool {
        return qtc.KACL_SetOwnerPermissions(@ptrCast(self), @intCast(ownerPermissions));
    }

    /// [Qt documentation](https://api.kde.org/kacl.html#owningGroupPermissions)
    ///
    /// ``` self: QtC.KACL ```
    pub fn OwningGroupPermissions(self: ?*anyopaque) u16 {
        return qtc.KACL_OwningGroupPermissions(@ptrCast(self));
    }

    /// [Qt documentation](https://api.kde.org/kacl.html#setOwningGroupPermissions)
    ///
    /// ``` self: QtC.KACL, owningGroupPermissions: u16 ```
    pub fn SetOwningGroupPermissions(self: ?*anyopaque, owningGroupPermissions: u16) bool {
        return qtc.KACL_SetOwningGroupPermissions(@ptrCast(self), @intCast(owningGroupPermissions));
    }

    /// [Qt documentation](https://api.kde.org/kacl.html#othersPermissions)
    ///
    /// ``` self: QtC.KACL ```
    pub fn OthersPermissions(self: ?*anyopaque) u16 {
        return qtc.KACL_OthersPermissions(@ptrCast(self));
    }

    /// [Qt documentation](https://api.kde.org/kacl.html#setOthersPermissions)
    ///
    /// ``` self: QtC.KACL, othersPermissions: u16 ```
    pub fn SetOthersPermissions(self: ?*anyopaque, othersPermissions: u16) bool {
        return qtc.KACL_SetOthersPermissions(@ptrCast(self), @intCast(othersPermissions));
    }

    /// [Qt documentation](https://api.kde.org/kacl.html#basePermissions)
    ///
    /// ``` self: QtC.KACL ```
    pub fn BasePermissions(self: ?*anyopaque) u32 {
        return @intCast(qtc.KACL_BasePermissions(@ptrCast(self)));
    }

    /// [Qt documentation](https://api.kde.org/kacl.html#isExtended)
    ///
    /// ``` self: QtC.KACL ```
    pub fn IsExtended(self: ?*anyopaque) bool {
        return qtc.KACL_IsExtended(@ptrCast(self));
    }

    /// [Qt documentation](https://api.kde.org/kacl.html#maskPermissions)
    ///
    /// ``` self: QtC.KACL, exists: *bool ```
    pub fn MaskPermissions(self: ?*anyopaque, exists: *bool) u16 {
        return qtc.KACL_MaskPermissions(@ptrCast(self), @ptrCast(exists));
    }

    /// [Qt documentation](https://api.kde.org/kacl.html#setMaskPermissions)
    ///
    /// ``` self: QtC.KACL, maskPermissions: u16 ```
    pub fn SetMaskPermissions(self: ?*anyopaque, maskPermissions: u16) bool {
        return qtc.KACL_SetMaskPermissions(@ptrCast(self), @intCast(maskPermissions));
    }

    /// [Qt documentation](https://api.kde.org/kacl.html#namedUserPermissions)
    ///
    /// ``` self: QtC.KACL, name: []const u8, exists: *bool ```
    pub fn NamedUserPermissions(self: ?*anyopaque, name: []const u8, exists: *bool) u16 {
        const name_str = qtc.libqt_string{
            .len = name.len,
            .data = name.ptr,
        };
        return qtc.KACL_NamedUserPermissions(@ptrCast(self), name_str, @ptrCast(exists));
    }

    /// [Qt documentation](https://api.kde.org/kacl.html#setNamedUserPermissions)
    ///
    /// ``` self: QtC.KACL, name: []const u8, param2: u16 ```
    pub fn SetNamedUserPermissions(self: ?*anyopaque, name: []const u8, param2: u16) bool {
        const name_str = qtc.libqt_string{
            .len = name.len,
            .data = name.ptr,
        };
        return qtc.KACL_SetNamedUserPermissions(@ptrCast(self), name_str, @intCast(param2));
    }

    /// [Qt documentation](https://api.kde.org/kacl.html#allUserPermissions)
    ///
    /// ``` self: QtC.KACL, allocator: std.mem.Allocator ```
    pub fn AllUserPermissions(self: ?*anyopaque, allocator: std.mem.Allocator) []struct_constu8_u16 {
        const _arr: qtc.libqt_list = qtc.KACL_AllUserPermissions(@ptrCast(self));
        defer {
            const _pair: [*]qtc.libqt_pair = @ptrCast(@alignCast(_arr.data));
            for (0.._arr.len) |i| {
                qtc.libqt_free(_pair[i].first);
                qtc.libqt_free(_pair[i].second);
            }
            qtc.libqt_free(_arr.data);
        }
        const _ret = allocator.alloc(struct_constu8_u16, _arr.len) catch @panic("kacl.AllUserPermissions: Memory allocation failed");
        const _data: [*]struct_constu8_u16 = @ptrCast(@alignCast(_arr.data));
        @memcpy(_ret, _data[0.._arr.len]);
        return _ret;
    }

    /// [Qt documentation](https://api.kde.org/kacl.html#namedGroupPermissions)
    ///
    /// ``` self: QtC.KACL, name: []const u8, exists: *bool ```
    pub fn NamedGroupPermissions(self: ?*anyopaque, name: []const u8, exists: *bool) u16 {
        const name_str = qtc.libqt_string{
            .len = name.len,
            .data = name.ptr,
        };
        return qtc.KACL_NamedGroupPermissions(@ptrCast(self), name_str, @ptrCast(exists));
    }

    /// [Qt documentation](https://api.kde.org/kacl.html#setNamedGroupPermissions)
    ///
    /// ``` self: QtC.KACL, name: []const u8, param2: u16 ```
    pub fn SetNamedGroupPermissions(self: ?*anyopaque, name: []const u8, param2: u16) bool {
        const name_str = qtc.libqt_string{
            .len = name.len,
            .data = name.ptr,
        };
        return qtc.KACL_SetNamedGroupPermissions(@ptrCast(self), name_str, @intCast(param2));
    }

    /// [Qt documentation](https://api.kde.org/kacl.html#allGroupPermissions)
    ///
    /// ``` self: QtC.KACL, allocator: std.mem.Allocator ```
    pub fn AllGroupPermissions(self: ?*anyopaque, allocator: std.mem.Allocator) []struct_constu8_u16 {
        const _arr: qtc.libqt_list = qtc.KACL_AllGroupPermissions(@ptrCast(self));
        defer {
            const _pair: [*]qtc.libqt_pair = @ptrCast(@alignCast(_arr.data));
            for (0.._arr.len) |i| {
                qtc.libqt_free(_pair[i].first);
                qtc.libqt_free(_pair[i].second);
            }
            qtc.libqt_free(_arr.data);
        }
        const _ret = allocator.alloc(struct_constu8_u16, _arr.len) catch @panic("kacl.AllGroupPermissions: Memory allocation failed");
        const _data: [*]struct_constu8_u16 = @ptrCast(@alignCast(_arr.data));
        @memcpy(_ret, _data[0.._arr.len]);
        return _ret;
    }

    /// [Qt documentation](https://api.kde.org/kacl.html#setACL)
    ///
    /// ``` self: QtC.KACL, aclStr: []const u8 ```
    pub fn SetACL(self: ?*anyopaque, aclStr: []const u8) bool {
        const aclStr_str = qtc.libqt_string{
            .len = aclStr.len,
            .data = aclStr.ptr,
        };
        return qtc.KACL_SetACL(@ptrCast(self), aclStr_str);
    }

    /// [Qt documentation](https://api.kde.org/kacl.html#asString)
    ///
    /// ``` self: QtC.KACL, allocator: std.mem.Allocator ```
    pub fn AsString(self: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.KACL_AsString(@ptrCast(self));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("kacl.AsString: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// [Qt documentation](https://api.kde.org/kacl.html#virtual_hook)
    ///
    /// ``` self: QtC.KACL, id: i32, data: ?*anyopaque ```
    pub fn VirtualHook(self: ?*anyopaque, id: i32, data: ?*anyopaque) void {
        qtc.KACL_VirtualHook(@ptrCast(self), @intCast(id), data);
    }

    /// [Qt documentation](https://api.kde.org/kacl.html#virtual_hook)
    ///
    /// Allows for overriding the related default method
    ///
    /// ``` self: QtC.KACL, callback: *const fn (self: QtC.KACL, id: i32, data: ?*anyopaque) callconv(.c) void ```
    pub fn OnVirtualHook(self: ?*anyopaque, callback: *const fn (?*anyopaque, i32, ?*anyopaque) callconv(.c) void) void {
        qtc.KACL_OnVirtualHook(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// [Qt documentation](https://api.kde.org/kacl.html#virtual_hook)
    ///
    /// Base class method implementation
    ///
    /// ``` self: QtC.KACL, id: i32, data: ?*anyopaque ```
    pub fn QBaseVirtualHook(self: ?*anyopaque, id: i32, data: ?*anyopaque) void {
        qtc.KACL_QBaseVirtualHook(@ptrCast(self), @intCast(id), data);
    }

    /// [Qt documentation](https://api.kde.org/kacl.html#dtor.KACL)
    ///
    /// Delete this object from C++ memory.
    ///
    /// ``` self: QtC.KACL ```
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.KACL_Delete(@ptrCast(self));
    }
};
